File: 2583-kth-largest-sum-in-a-binary-tree/2583-kth-largest-sum-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Pair{
    TreeNode root;
    int level;
    Pair(TreeNode root, int level){
      this.root=root;
      this.level=level;
    }
  }
class Solution {
    public long kthLargestLevelSum(TreeNode root, int k) {
        Queue<Pair> q=new LinkedList<>();

        q.add(new Pair(root, 0));
        ArrayList<Long> res=new ArrayList<>();
        int maxlevel=0;
        
        while(!q.isEmpty()){
          int size=q.size();
          int l=0;
            maxlevel++;
          long sum=0;
          for(int i=0;i<size;i++){
            Pair p=q.poll();
            l=p.level;
            //maxlevel=p.level;
            TreeNode curr=p.root;
            sum+=curr.val;
            if(curr.left!=null) q.add(new Pair(curr.left, l+1));
            if(curr.right!=null) q.add(new Pair(curr.right, l+1));

          }
          res.add(sum);


        }
        Collections.sort(res, Collections.reverseOrder());
        if(maxlevel<k) return -1;
        return res.get(k-1);
    }
}

File: 2583-kth-largest-sum-in-a-binary-tree/2583-kth-largest-sum-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long long kthLargestLevelSum(TreeNode* root, int k) {
          if(!root){
            return {};
        }
        vector<long long> ans;
        vector<int> level;
        queue<TreeNode*> q;
        level.push_back(root->val);
        ans.push_back(root -> val);
        level.clear();
        q.push(root);
        
        
        
        while(!q.empty()){
            int n = q.size();
            for(int i=0;i<n;i++){
                TreeNode* tmp = q.front();
                q.pop();
                
                if(tmp -> left){
                    q.push(tmp -> left);
                    level.push_back(tmp -> left -> val);
                }
                
                if(tmp -> right){
                    q.push(tmp ->  right);
                    level.push_back(tmp -> right -> val);
                }
            }
            
            if(!level.empty()){
                long long num=0;
                for(auto i : level)
                    num += i;
                    
                ans.push_back(num);
                level.clear();

            }
        }
        
         sort(ans.begin(),ans.end(), greater <>());
        
        int n = ans.size();
        
        if(n>=k)
            return ans[k-1];
        else
            return -1;
        
       
    }
};